clash-lib/src/Clash/Normalize/Transformations.hs

```diff
@@ -494,150 +490,150 @@ inlineNonRep _ e = return e
-        -- Use the original inScopeSet 'is0' here, not the extended inScopeSet
-        -- 'is1', otherwise we'd make the "caseCon1" substitution substitute
-        -- free variables that were shadowed by the pattern!
-        let subst = extendTvSubstList (mkSubst is0)
-                  $ zip tvs (drop (length (dcUnivTyVars dc)) (Either.rights args))
-        changed (substTm "caseCon1" subst e')
-      _ -> case alts of
-             ((DefaultPat,e):_) -> changed e
-             _ -> changed (undefinedTm ty)
-  where
-    equalCon dc (DataPat dc' _ _) = dcTag dc == dcTag dc'
-    equalCon _  _                 = False
-
-    newBinder (isN0,substN) (x,arg)
-      | isWorkFree arg
-      = ((isN0,(x,arg):substN),Nothing)
-      | otherwise
-      = let x'   = uniqAway isN0 x
-            isN1 = extendInScopeSet isN0 x'
-        in  ((isN1,(x,Var x'):substN),Just (x',arg))
-
-caseCon _ c@(Case (stripTicks -> Literal l) _ alts) = case List.find (equalLit . fst) alts of
-    Just (LitPat _,e) -> changed e
-    _ -> matchLiteralContructor c l alts
-  where
-    equalLit (LitPat l')     = l == l'
-    equalLit _               = False
-
-caseCon ctx@(TransformContext is0 _) e@(Case subj ty alts)
-  | (Prim _,_) <- collectArgs subj = do
-    reprs <- Lens.view customReprs
-    tcm <- Lens.view tcCache
-    bndrs <- Lens.use bindings
-    (primEval, primUnwind) <- Lens.view evaluator
-    ids <- Lens.use uniqSupply
-    let (ids1,ids2) = splitSupply ids
-    uniqSupply Lens..= ids2
-    gh <- Lens.use globalHeap
-    lvl <- Lens.view dbgLevel
-    case whnf' primEval primUnwind bndrs tcm gh ids1 is0 True subj of
-      (gh',ph',v) -> do
-        globalHeap Lens..= gh'
-        bindPureHeap ctx tcm ph' $ \ctx' -> case stripTicks v of
-          Literal l -> caseCon ctx' (Case (Literal l) ty alts)
-          subj' -> case collectArgsTicks subj' of
-            (Data _,_,_) -> caseCon ctx' (Case subj' ty alts)
+caseCon ctx@(TransformContext is0 _) e@(Case subj ty alts) = case collectArgsTicks subj of
+  -- The subject is an applied data constructor
+  (Data dc, args, ticks) -> case List.find (equalCon . fst) alts of
+    Just (DataPat _ tvs xs, altE) -> do
+      let is1 = extendInScopeSetList (extendInScopeSetList is0 tvs) xs
+      let fvs = Lens.foldMapOf freeLocalIds unitVarSet altE
+          (binds,_) = List.partition ((`elemVarSet` fvs) . fst)
+                    $ zip xs (Either.lefts args)
+          binds1 = map (second (`mkTicks` ticks)) binds
+          altE1 = case binds1 of
+            [] -> altE
+            _  ->
+              -- See Note [CaseCon deshadow]
+              let ((is3,substIds),binds2) =
+                    List.mapAccumL newBinder (is1,[]) binds1
+                  subst = extendIdSubstList (mkSubst is3) substIds
+                  body  = substTm "caseCon0" subst altE
+              in  case Maybe.catMaybes binds2 of
+                    []     -> body
+                    binds3 -> Letrec binds3 body
```

Comment:
No change requested, but I've found a good way of formatting `let` bindings without indenting "too much" is:

```
let
  a = 123
  b = 456
in
  a + b
```